### PR TITLE
Adds the rat king

### PR DESCRIPTION
**Pull request by Alberyk**
Ports the rat king from apollo, as an admin tool for now.

The rat king can collect mouse as he walks by, getting stronger, and unlocking some skills, it also can speak to all mouse in the world via a special skill.

### CODE REVIEW

File: code/modules/mob/living/simple_animal/friendly/ratking.dm [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		R << message
+
+// Since there's no proc for the current object to check crossed objects
+/mob/living/simple_animal/mouse/Crossed(atom/movable/O)
+	..()
+
+	if( !health )
+		return
+
+	if( istype(O,/mob/living/simple_animal/mouse/king))
```
Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: Why can't this be defined as `/mob/living/simple_animal/mouse/king/Crossed()`?

File: code/modules/mob/living/simple_animal/friendly/ratking.dm [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+		var/mob/living/simple_animal/mouse/king/K = O
+		if( !K.health )
+			return
+
+		src.visible_message("<span class='warning'>[src] joins the [K.swarm_name] of \the [K]</span>", \
+							"<span class='notice'>We join our brethren in \the [K.swarm_name]. Long live \the [K].<span>")
+		K.absorb(src)
+
+/mob/living/simple_animal/mouse/king
```
Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: Why is this in friendly?

File: code/modules/mob/living/simple_animal/friendly/ratking.dm [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+	if(!(R in rats))
+		R.Move(src)
+		rats += R
+
+	if( update )
+		update()
+
+/mob/living/simple_animal/mouse/king/proc/eject(var/mob/living/simple_animal/mouse/R, var/update = 1)
+	if(R in rats)
+		R.Move(get_turf(src))
```
Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: `forceMove`.

File: code/modules/mob/living/simple_animal/friendly/ratking.dm [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+	src.visible_message("<span class='danger'>\The [usr] consumed the body of \the [target]!</span>")
+	target.gib()
+	health = maxHealth
+	last_special = world.time + 100
+	return
+
+/mob/living/simple_animal/mouse/king/proc/absorb(var/mob/living/simple_animal/mouse/R, var/update = 1)
+	if(!(R in rats))
+		R.Move(src)
```
Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: `forceMove`.

File: code/modules/mob/living/simple_animal/friendly/ratking.dm [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+	while(rats.len)
+		eject(rats[1], 1)
+
+	..()
+
+/mob/living/simple_animal/mouse/king/Move()
+	..()
+
+	for(var/mob/living/simple_animal/mouse/rat in rats)
+		rat.dir = src.dir
```
Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: Why is this necessary? The rats are located inside the rat king mob, they shouldn't be visible.

File: code/modules/mob/living/simple_animal/friendly/ratking.dm (hunk 2 above)
Comment (Alberyk) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: For the same reason mouse are as well.

Comment (Lohikar) [inline code/modules/mob/living/simple_animal/friendly/ratking.dm]: It doesn't seem that friendly, but I guess it makes sense as a mouse child-type.

File: code/modules/mob/living/simple_animal/friendly/mouse.dm [hunk 6]
```diff
@@ -1,3 +1,5 @@
+var/global/list/world_mouses = list()
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/friendly/mouse.dm]: ...Jesus christ, is this really necessary?

Also, there's a file with all of the global lists. It should be there.

Comment (Alberyk) [inline code/modules/mob/living/simple_animal/friendly/mouse.dm]: Done.

**Comment by Alberyk** [comment]
Also done.

File: code/modules/mob/living/simple_animal/friendly/mouse.dm [hunk 7]
```diff
@@ -80,7 +79,12 @@
 	else
 		if ((world.time - timeofdeath) > decompose_time)
 			dust()
-
+			
+/mob/living/simple_animal/mouse/Destroy()
+	..()
+	
+	world_mouses -= src
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/friendly/mouse.dm]: As per code standards:
```
/mob/living/simple_animal/mouse/Destroy()
	world_mouses -= src

	return ..()
```

**Comment by Alberyk** [comment]
And fixed.

**Comment by Lohikar** [comment]
Since SMC's merged now, can you please move the global mouse list into `SSmobs`?

File: code/global.dm [hunk 8]
```diff
@@ -16,6 +16,7 @@ var/global/list/med_hud_users            = list() // List of all entities using
 var/global/list/sec_hud_users            = list() // List of all entities using a security HUD.
 var/global/list/hud_icon_reference       = list()
 var/global/list/janitorial_supplies      = list()	// List of all the janitorial supplies on the map that the PDA cart may be tracking.
+var/global/list/world_mouses 		 = list() //List of all mice in the world
```
Comment (Lohikar) [inline code/global.dm]: Move this into `SSmob` pls.

Comment (Lohikar) [inline code/global.dm]: Actually, I forgot to actually give SSmobs a SS global, so never mind.

Comment (skull132) [inline code/global.dm]: Stop memein'. :ree:
